R/fullVALIDATIONREMIND.R

```diff
@@ -13,46 +13,57 @@
+  # of the `regionmapping` and `extramappings` currently configured.
 
-  #-------------- warn if there are duplicate column names --------------------------------------------
-
-  # read region mappings
-  mappings <- c(getConfig("regionmapping"), getConfig("extramappings"))
-  rel <- NULL
-  for (r in seq_len(length(mappings))) {
-    # read mapping file
-    new <- colnames(toolGetMapping(mappings[r], type = "regional"))
-    # ignore "X" and "CountryCode"
-    new <- new[!new %in% c("X", "CountryCode")]
-    # check if columns of current mapping exist in any mapping read before
-    if (any(new %in% rel)) warning("The following column(s) from ", mappings[r],
-                                   " exist in another mapping an will be ignored: ",
-                                   paste(new[new %in% rel], collapse = ", "))
-    # append new columns removing duplicates
-    rel <- unique(c(rel, new))
-  }
-
-  #-------------- use columns from all regionmappings for aggregation ---------------------------------
-
-  columnsForAggregation <- gsub("RegionCode", "region", paste0(c(rel, "global"), collapse = "+"))
+  rel <- 'global'   # always compute global aggregate
+  for (mapping in c(getConfig("regionmapping"), getConfig("extramappings"))) {
+    columns <- setdiff(colnames(toolGetMapping(mapping, 'regional')),
+                       c('X', 'CountryCode'))
 
-  #-------------- historical data ---------------------------------------------------------------------
+    if (any(columns %in% rel))
+      warning("The following column(s) from ", mapping,
+              " exist in another mapping an will be ignored: ",
+              paste(columns[columns %in% rel], collapse = ", "))
 
-  valfile <- "historical.mif"
+    rel <- unique(c(rel, columns))
+  }
 
-  calcOutput("Historical", round = 5,  file = valfile, aggregate = columnsForAggregation,
-             append = FALSE, na_warning = FALSE, try = TRUE)
+  columnsForAggregation <- gsub("RegionCode", "region",
+                                paste(rel, collapse = "+"))
 
-  calcOutput(type = "IEA_ETP", aggregate = "global", file = valfile,
-             append = TRUE, na_warning = FALSE, try = TRUE, isValidation = TRUE)
+  # default arguments ----
+  # default arguments used for calling all subsequent `calc` functions
+  default_arguments <- list(
+    file = 'historical.mif',
+    append = TRUE,
+    warnNA = FALSE,
+    aggregate = columnsForAggregation,
+    try = TRUE)
 
-  calcOutput(type = "IEA_ETP", aggregate = "region", file = valfile,
-             append = TRUE, na_warning = FALSE, try = TRUE, isValidation = TRUE)
+  # list `calc` functions to call ----
+  # list of all `calc` functions to call, as well as their specific arguments
+  # arguments can be empty (`list()`), and can overwrite default arguments if
+  # needed
+  historical_data_calls <- list(
+    'Historical'   = list(round = 5),
+    'IEA_ETP'      = list(isValidation = TRUE),
```

Comment:
IEA_ETP and WEO_2021 are special cases, where setting `columnsForAggregation` won't produce the expected data. You can only pass `region`and `global` in these two specific cases right now. 

Background: madrat does not have a way to retrieve different variable sets depending on the regional aggregation (IEA ETP and WEO have some variables we can only use on global level). So we must work with dirty hacks. (The alternative is to create two models, e.g IEA WEO regional and IEA WEO global with different variable sets, which clutters compareScenarios).

Reply:
Yes, it is not good, but so far we haven't found a better way to select specific variables depending on intended regional granularity we want to use it for. If it breaks one day, we will just have to either include or remove these two variables no matter the aggregation

Reply:
> You are excluding these variables only if aggregation equals 'region', when in fact you want to exclude them whenever aggregation is anything other than 'global'.

The problem is, aggregate could also be `region+global+FOO+BAR`.  It is passed like this into the calc function from what i can see.

Reply:
> If it is a general problem that our aggregation is batshit crazy less then ideal, I would argue for a blacklist/whitelist mechanism to filter the resulting .mif file.

Yes, this would probably then have to be done separately in the libraries working with historical.mif, especially remind2's compareScenarios.